File: day-02/solution.exs
games = File.read!("./input.txt")
  |> String.splitter("\n", trim: true)

defmodule Day02 do

  @game_map %{"X" => "A", "Y" => "B", "Z" => "C"}

  def getValue(x) do
    case x do
      "A" -> 1
      "B" -> 2
      _ -> 3
    end
  end

  def games(x) do
    Stream.map(x, &String.split(&1, " "))
    |> Enum.map(fn [o, u] -> [o, @game_map[u]] end)
    |> Enum.map(&scorer/1)
  end

  def scorer([o, u]) do
    val = getValue(u)
    cond do
      u == o -> 3 + val
      (o == "A" && u == "B") ||(o == "B" && u == "C") ||(o == "C" && u == "A") -> 6 + val
      true -> val
    end
  end

  def getWinner(x) do
    (
      case x do
        "A" -> "B"
        "B" -> "C"
        _ -> "A"
      end
    )
    |> getValue
  end

  def getLoser(x) do
    (
      case x do
        "C" -> "B"
        "A" -> "C"
        _ -> "A"
      end
    )
    |> getValue
  end

  def scorer2(x) do
    [o, result] = String.split(x, " ")
    case result do
      "Y" -> 3 + getValue(o)
      "Z" -> 6 + getWinner(o)
      _ -> getLoser(o)
    end
  end
end


part01 = games
  |> Day02.games()
  |> Enum.sum()

IO.puts(part01)

part02 = games
  |> Enum.map(&Day02.scorer2/1)
  |> Enum.sum()

IO.puts(part02)

File: day-02/solution.js
const { readFileSync } = require("fs");

const gameMap = new Map();
gameMap.set("X", "A");
gameMap.set("Y", "B");
gameMap.set("Z", "C");

const getValue = (val) => (val === "A" ? 1 : val === "B" ? 2 : 3);

const games = readFileSync("./input.txt", "utf-8").split("\n");

const part01 = games
  .map((game) => game.split(" "))
  .map(([o, u]) => [o, gameMap.get(u)])
  .reduce((acc, cv) => {
    const [op, u] = cv;

    if (op === u) {
      acc += 3;
    } else if (
      (op === "A" && u === "B") ||
      (op === "B" && u === "C") ||
      (op === "C" && u === "A")
    ) {
      acc += 6;
    }

    acc += getValue(u);
    return acc;
  }, 0);

console.log(part01);

const part02 = games
  .map((game) => game.split(" "))
  .reduce((acc, cv) => {
    const [op, result] = cv;

    if (result === "Y") {
      acc += 3;
      acc += getValue(op);
    } else if (result === "Z") {
      acc += 6;
      acc += getValue(op === "A" ? "B" : op === "B" ? "C" : "A");
    } else {
      acc += getValue(op === "A" ? "C" : op === "B" ? "A" : "B");
    }
    return acc;
  }, 0);

console.log(part02);
